Various tests improvements

paolochiodi opened the pull request:
This PR includes:
- make tests self contained, thus order independent
- expand tests to fully cover specs as per https://github.com/rjrodger/seneca/blob/master/doc/data-store.md#cmdremove and discussions in the following issues: https://github.com/rjrodger/seneca-store-test/issues/13, https://github.com/rjrodger/seneca-store-test/issues/14, https://github.com/rjrodger/seneca-store-test/issues/15, https://github.com/rjrodger/seneca-store-test/issues/16 and https://github.com/rjrodger/seneca-store-test/issues/17
- adhere to "standard" style adopted for all seneca related repos
- bring dependencies up to date

This PR does not include:
- updates to the README
- version bump
- wrapped errors as discussed in https://github.com/rjrodger/seneca-store-test/issues/14 (I'm working on an additional PR to bring that, but would like to get this merged first)

Notes:
- instead of relying on test.sh to install seneca I added it to devDependencies. I'm not 100% sure it's the right place
- this will require an update to how tests are run on the stores, since it now requires a "senecaMerge" option. That should be an instance of seneca with the store configured with the merge option set to false
- tested with seneca from 0.5 to 0.7, node 0.10, 0.12, 4, 4.1
- mem-store included with seneca fails some of the tests

paolochiodi commented:
Please note that travis checks fail because the men-store used for the tests doesn't fully follow standards

geek commented on .travis.yml:
```diff
@@ -1,4 +1,6 @@
 language: node_js
 node_js:
-  - "0.11"
+  - "4.1"
```
Can use `"4"` instead

paolochiodi replied:
do you think I should add "4" and also leave 4.1 and 4.0 since 4.2 has been released?

geek replied:
I would make it "4" to test against the current 4.x release.

geek commented on package.json:
```diff
@@ -21,19 +21,23 @@
   ],
+  "dependencies": {
+    "async": "^1.4.0",
+    "chai": "^3.3.0",
```
should move into devDependencies

paolochiodi replied:
I'm not sure here. These are dependencies needed to run this package, not to test it.
That's because this package is kind of _meta_ in that it is used to run tests. seneca-store-test will be among the devDependencies of other stores, but async and chai are required to make this work.
In fact I actually think I should move even lab here.

mcdonnelldean replied:
Regular dependencies is fine here. The package is a test package so these are its direct dependencies.

mcdonnelldean commented:
@paolochiodi Can you look at the failing tests? There are a couple of them.

mcdonnelldean commented on package.json:
```diff
@@ -4,7 +4,7 @@
   "main": "store-test.js",
   "scripts": {
-    "test": "./test.sh"
+    "test": "lab test/*.test.js -r console -v -L"
```
You might want to set the timeout here to 3000 we have found travis can be slow at times and cause tests to fail due to overrun. 3000 seems to be the sweet spot.

mcdonnelldean commented on store-test.js:
```diff
@@ -62,426 +54,997 @@ function verify (cb, tests) {
+        }))
+      })
 
-    })
```
Do we need two tests here? Are both filter tests the same?

paolochiodi replied:
I just wanted to also have a test for a filter that doesn't find any entity.
Unfortunately, having only two records this test may result in false positive (e.g.: if there's a bug ignoring the second half of the query `p2: 'z2'`.
The following test will test for that.

paolochiodi commented:
@mcdonnelldean tests on this repot aren't real tests
They actually use the tests in this module to test the mem-store built in seneca (which I think is considered a "reference implementation". 

I've introduced some new test to better cover the data entities specification, and some them are failing because even men-store doesn't fully and correctly implement that. In order to have all tests green we should amend the mem-store

mcdonnelldean commented:
@paolochiodi super, as discussed let's fix up mem store while we wait for @rjrodger to review.

geek commented:
@paolochiodi this change looks good... I would like to get the changes to mem-store in first so that we can have all green tests.  I went ahead and merged in #22, so you will need to pull down those changes and merge :/

paolochiodi commented:
I've rebased to incorporate latest changes in master branch.
I've also updated seneca and seneca-men-store to make all tests pass (relevant PR here https://github.com/senecajs/seneca/pull/217 and here https://github.com/senecajs/seneca-mem-store/pull/36)

Releasing these will require some coordination and some work on package.json to highlight that these spec changes:
1. break existing spec and stores
2. require latest versions of seneca (the ones including https://github.com/senecajs/seneca/pull/217)

paolochiodi commented:
Rebased to bring this up-to-date with master

mirceaalexandru commented on store-test.js:
```diff
@@ -61,250 +55,645 @@ function verify (cb, tests) {
-      scratch.foo2.p2 = 'v2'
+      it('should allow to not merge during update with merge$: false', function (done) {
+        var foo = si.make('foo')
+        foo.id = 'to-be-updated'
```
@paolochiodi I think is better not to define id values in these tests. Some DB engines have rules about ids - like MongoDB that uses an ObjectId. So in order to use these tests as a standard maybe we should not have id values defined in these tests.

The id can be retrieved from callback of the save action.

paolochiodi replied:
I guess this also relates to https://github.com/senecajs/seneca-shard-store/issues/11
On a general note I'm ok to rewrite tests in a way that `createEntities` will store auto-generated ids for later uses (i.e.: when we want to update the entities, like this case).

I begin to feel that at this point may be better to remove the possibility to specify the id on create from the specs, if many of the DB engines can't support them. What do you think?

mirceaalexandru replied:
I am the supporter of the idea that we should have something like this:
- define a set of functions that all stores will try to follow - "standard" store functions.
- support in the stores as many as possible "standard" functions supported by the DB engines. Here there is also this function of allowing to specify the id or the entity to be saved if that is supported.
- clearly specify in the documentation when a certain store do not have support for a "standard" function.

So, from my point of view, if Riak do not support list$ is not a problem, but we should clearly specify this in its documentation. The developers that will use that store probably know already this thing. 

I also think that the benefit of having a functional store with all necessary tools offered by a specific DB engine is greater than limiting all stores to the smallest set of common functions offered by all DB engines.

Shard-store can also have a list of specific functions that should not be used by developers if they want to have it run correctly - this should not limit the actual functions implemented in the used stores.

paolochiodi replied:
I agree on the idea of having a basic set of standard functions, but I think all the stores _must_ support that minimal set.

The entity framework is quite minimal and doesn't add much to the db driver beside a common api, I don't see the point of having a store for a db that can't support that api: sometimes may be better to directly go with the driver or a different type of abstraction (e.g: https://github.com/AdrianRossouw/seneca-knex-store)

Maybe @rjrodger and @geek can weight in their opinion?

Also let me know how to proceed with this PR, if I need to split the tests into different "levels" of support
